### DISCUSSION TITLE
Unable to run phase-field fracture simulations using spectral decomposition of strain energy

### DISCUSSION THREAD

**Discussion by MartimSalgado** [idaholab/moose#18265] [Q&A Modules: Phase field] [answered]
Hi,

I've created a model for a single edge notched tension test loaded in displacement control in moose using phase-field coupled with tensor mechanics and a volumetric deviatoric decomposition of strain energy. However, when I try to change the decomposition type to "spectral", the solver always fails to converge at the early steps of the simulation. I've already tried to use different petsc options, mesh, displacement increment, nothing works. I'm also running 2 versions: monolithic and staggered schemes (one single app and using multiapp). I'm sending attached the monolithic version that works with volumetric deviatoric.

Does anyone have any tips on how to get this to work with spectral decomposition? Thank you!
[sent.zip](https://github.com/idaholab/moose/files/6768661/sent.zip)

**Comment by hugary1995** [comment]
Hi, thanks for reporting this. Could you convert this into an issue? I will look into this.

**Comment by jiangwen84** [comment, marked as answer]
@MartimSalgado  Couple of things: 

- I have to comment out some blocks that use your own staff.  
- Since you use VI solver, you should set `use_snes_vi_solver = true` 
- You used 8 levels of refinement which seems to be too many. I suggest starting with a uniform mesh to identify the convergence issue. 
- You have two identical BC blocks, `top_x` and `bottom_x`
- @hugary1995 and I found an error of jacobian calculation in `ComputeLinearElasticPFFractureStress::computeStrainSpectral` It should not affect the convergence much because of using PJFNK, but we still need to fix it as soon as we can.

**Reply by ghost** [reply]
Hello moose developers,
I am new moose user and interested in developing models for phase field fracture in polycrystals. I have been trying to reproduce 
pure shear tests published in many articles using spectral decomposition. First I considered newton method and figured out that there was an issue because of the jacobian so I implemented a new jacobian based on the projector of the strain_old  (or stress_old) which is normally stable. However even with this I could not have convergence. Could anyone help us in solving this ?
Thank you in advance.

**Reply by jiangwen84** [reply]
> Hello moose developers,
> 
> I am new moose user and interested in developing models for phase field fracture in polycrystals. I have been trying to reproduce 
> 
> pure shear tests published in many articles using spectral decomposition. First I considered newton method and figured out that there was an issue because of the jacobian so I implemented a new jacobian based on the projector of the strain_old  (or stress_old) which is normally stable. However even with this I could not have convergence. Could anyone help us in solving this ?
> 
> Thank you in advance.
> 
> 

What was the issue with the oringal code?  The Jabobian should be correct.

**Reply by ghost** [reply]
Thanks for your reply.
The jacobian is correct, but we think that the spectral method would not converge with a  vacillating Jacobian due to a null eigen value in a pure shear test.  That is why we have considered the "old" Projector instead of the current when computing the Jacobian, and even with this, we could not have convergence.
regards,

**Reply by GiudGiud** [reply]
Hello

You should test the jacobian using our automatic tools to determine that
See this page:
https://mooseframework.inl.gov/help/development/analyze_jacobian.html
Guillaume

**Reply by jiangwen84** [reply]
> Thanks for your reply.
> The jacobian is correct, but we think that the spectral method would not converge with a vacillating Jacobian due to a null eigen value in a pure shear test. That is why we have considered the "old" Projector instead of the current when computing the Jacobian, and even with this, we could not have convergence.
> regards,

I do not remember we have much convergence trouble with the pure shear test. Does it converge badly on linear or nonlinear iterations?

Could you paste your "executioner" block here?

**Reply by ghost** [reply]
We have tried many execution scenarios, the last one is :
[Executioner]
[./TimeStepper]
    type = IterationAdaptiveDT
    dt =1.0e-8  # Initial time step.  In this simulation it changes.
    optimal_iterations = 8 # Time step will adapt to maintain this number of nonlinear iterations
    growth_factor = 1.5
  [../]
  type = Transient
  #solve_type =newton #PJFNK
petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -snes_type'
  petsc_options_value = 'lu       superlu_dist                  vinewtonrsls'
  automatic_scaling = true
  line_search = 'none'
  nl_rel_tol = 1e-3
  l_tol = 1.0e-4
  l_max_its = 200
  nl_max_its = 14
  dtmax = 3.0e-2
  dtmin = 1e-11
 end_time = 20.0
[]

**Reply by hugary1995** [reply]
See this app as an example: https://github.com/hugary1995/raccoon/tree/devel/tutorials/mode2_brittle_fracture

> The jacobian is correct, but we think that the spectral method would not converge with a vacillating Jacobian due to a null eigen value in a pure shear test.

I also don't quite understand this statement.

**Reply by ghost** [reply]
Of course I did not face problems with non pre-cracked meshs. However with a pre-cracked one, like in many articles, it never converged. The pre-crack could be either d=1 or a real mesh cut.
regards

**Reply by hugary1995** [reply]
> However with a pre-cracked one, like in many articles, it never converged.

But many articles show results generated with MOOSE...

**Reply by ghost** [reply]
That is why we are asking how these results have been obtained

**Reply by hugary1995** [reply]
Fair enough. All I can say is the implementation in MOOSE is correct, and we are able to reproduce the pure shear benchmark. Please either set up a branch containing necessary files to reproduce the convergence issue (then someone may help you take a look) or contact the corresponding authors of the articles you mentioned.

**Reply by selarem** [reply]
I could say that I have been facing the same problem with pure shear test and never got convergence with spectral decomposition.

**Reply by selarem** [reply]
with a pre-cracked mesh.

**Reply by hugary1995** [reply]
I've already provided an example: https://github.com/hugary1995/raccoon/tree/devel/tutorials/mode2_brittle_fracture
Saying you never got it to work without providing any detail doesn't help us advance this discussion.

**Reply by selarem** [reply]
Thank you for the example, I will try it but it looks like that it is not a pure shearing test. I think that in a pure shear test the disp_y has to be zero everywhere.

**Reply by hugary1995** [reply]
Great, we are getting somewhere :)

So I may have totally misunderstood your "pure shear" test. By "disp_y has to be zero everywhere", do you mean you also want to set disp_y on the left and right boundaries, or do you mean you literally want to set disp_y _everywhere_ on the domain, i.e., on all the nodes?

**Reply by hugary1995** [reply]
> disp_y = 0 everywhere on the external boundaries: left,right, top and bottom.

In that case, I believe the example I linked should also work if you modify the boundary conditions to match what you described, i.e., there should not be any convergence issue.

**Reply by selarem** [reply]
The example in raccoonn is working and gives the crack path as per the published articles.
However, I prefer running this in moose where things are easier for me to understand. 
I have tried with these material and mesh parameters (as in the article by Miehe et al. in IJNME, 2010)  many times with no success.

**Reply by selarem** [reply]
an example of non convergence problem.
https://drive.google.com/drive/folders/1nAjF0WhgQez2ZiiGKQVvbIRZkEiQc-pp?usp=sharing
Here we use the moose files. We have also tried with the old projectors.
regards,

**Reply by jiangwen84** [reply]
I take a quick scan of you input file, your ```nl_rel_tol = 1e-3``` and ```nl_abs_tol = 1e-6```  is too loose, considering you are using ```automatic_scaling```. 

Let me try to run your input file and will let you know more.

**Reply by selarem** [reply]
Any solution for this issue ? I have the same problem even with nl_rel_tol = 1e-5; even with old projectors.

**Reply by selarem** [reply]
Always,  convergence issue ! We have tried everything and it took too much time for Ph.D. students trying to reproduce results from 
litterature !

**Reply by jiangwen84** [reply]
@selarem Who is the advisor of the PhD students? I feel that it is their advisor's responsibility to help their Ph.D to figure out the  convergence issues. The codes are open source, and you can see every pieces of the codes. If you find any errors, please feel free to fix them. 

I haven't looked at your problem closely. It is not a small problem to run, so please be patient. One thing you can consider is to use a staggered approach.  Typically, that works much better than the fully coupled approach.

**Reply by selarem** [reply]
You are right. We have used the old projector, and corrected the stress computation but it does not converge in a pure shear test.
This test is like a benchmark since it is in all reference papers; that is why we start from there. In a staggered way it works fine with our "home" codes, in raccoon and in moose. But we want to understand why.

**Reply by jiangwen84** [reply]
@selarem Thanks for your update. Let me take a closer look in the next few days.

**Comment by selarem** [comment]
disp_y = 0 everywhere on the external boundaries: left,right, top and
bottom.
